GVFS/GVFS.Common/ModifiedPathsDatabase.cs

```diff
@@ -69,44 +69,86 @@ public bool TryAdd(string path, bool isFolder, out bool isRetryable)
             string entry = this.NormalizeEntryString(path, isFolder);
             if (!this.modifiedPaths.Contains(entry))
             {
-                try
+                return this.TryAddOrRemove(entry, isFolder, out isRetryable, addEntry: true);
+            }
+
+            return true;
+        }
+
+        public bool TryRemove(string path, bool isFolder, out bool isRetryable)
+        {
+            isRetryable = true;
+            string entry = this.NormalizeEntryString(path, isFolder);
+            if (this.modifiedPaths.Contains(entry))
+            {
+                return this.TryAddOrRemove(entry, isFolder, out isRetryable, addEntry: false);
+            }
+
+            return true;
+        }
+
+        public void WriteAllEntriesAndFlush()
+        {
+            try
+            {
+                this.WriteAndReplaceDataFile(this.GenerateDataLines);
+            }
+            catch (Exception e)
+            {
+                throw new FileBasedCollectionException(e);
+            }
+        }
+
+        private IEnumerable<string> GenerateDataLines()
+        {
+            foreach (string entry in this.modifiedPaths)
+            {
+                yield return this.FormatAddLine(entry);
+            }
+        }
+
+        private bool TryAddOrRemove(string entry, bool isFolder, out bool isRetryable, bool addEntry)

```

Comment:
This seems like it should be two separate methods, since it does two non-overlapping actions based on the value of a bool parameter

Reply:
Yeah I went this way because there were essentially ~15 lines of duplicate code with 2 separate methods but this does feel a little hacky and not as straight forward, do I'll switch it back.